[analyzer] Introduce the check::BlockEntrance checker callback

balazs-benics-sonarsource opened the pull request:
Tranersing the CFG blocks of a function is a fundamental operation. Many C++ constructs can create splits in the control-flow, such as `if`, `for`, and similar control structures or ternary expressions, gnu conditionals, gotos, switches and possibly more.

Checkers should be able to get notifications about entering or leaving a CFG block of interest.

Note that in the ExplodedGraph there is always a BlockEntrance ProgramPoint right after the BlockEdge ProgramPoint. I considered naming this callback check::BlockEdge, but then that may leave the observer of the graph puzzled to see BlockEdge points followed more BlockEdge nodes describing the same CFG transition. This confusion could also apply to Bug Report Visitors too.

Because of this, I decided to hook BlockEntrance ProgramPoints instead. The same confusion applies here, but I find this still a better place TBH. There would only appear only one BlockEntrance ProgramPoint in the graph if no checkers modify the state or emit a bug report. Otherwise they modify some GDM (aka. State) thus create a new ExplodedNode with the same BlockEntrance ProgramPoint in the graph.

CPP-6484

balazs-benics-sonarsource committed:
[analyzer] Introduce the check::BlockEntrance

Tranersing the CFG blocks of a function is a fundamental operation.
Many C++ constructs can create splits in the control-flow, such as
`if`, `for`, and similar control structures or ternary expressions,
gnu conditionals, gotos, switches and possibly more.

Checkers should be able to get notifications about entering or leaving a
CFG block of interest.

Note that in the ExplodedGraph there is always a BlockEntrance
ProgramPoint right after the BlockEdge ProgramPoint.
I considered naming this callback check::BlockEdge, but then that may
leave the observer of the graph puzzled to see BlockEdge points followed
more BlockEdge nodes describing the same CFG transition.
This confusion could also apply to Bug Report Visitors too.

Because of this, I decided to hook BlockEntrance ProgramPoints instead.
The same confusion applies here, but I find this still a better place
TBH. There would only appear only one BlockEntrance ProgramPoint in the
graph if no checkers modify the state or emit a bug report.
Otherwise they modify some GDM (aka. State) thus create a new
ExplodedNode with the same BlockEntrance ProgramPoint in the graph.

CPP-6484

NagyDonat reviewed (commented):
I'll try to review this commit within a few days -- please ping me if I accidentally forget.

I have one immediate question about this area:

> Traversing the CFG blocks of a function is a fundamental operation. Many C++ constructs can create splits in the control-flow, such as if, for, and similar control structures or ternary expressions, gnu conditionals, gotos, switches and possibly more.

What is the relationship of this new callback with the `BranchCondition` callback which is evaluated in `ProcessBranch` where the checker splits the execution path into multiple branches? It would be nice if you could document the difference between these two callbacks.

Xazax-hun commented:
> What is the relationship of this new callback with the BranchCondition callback

+1, I am also interested to learn what is the main motivation behind the new callback. Do you have some example use cases in mind? I think users might get confused which one to use.

balazs-benics-sonarsource committed:
Clarify the relation between BranchCondition and BlockEntrance

balazs-benics-sonarsource committed:
Add an example unit test where both checkers are in action

balazs-benics-sonarsource commented:
> What is the relationship of this new callback with the `BranchCondition` callback which is evaluated in `ProcessBranch` where the checker splits the execution path into multiple branches? It would be nice if you could document the difference between these two callbacks.

Great question. They are for different purposes, but one should prefer `BranchCondition` over the strictly more generic `BlockEntrance` callback. State splits work as in any other checkers. If you split the state in both callback, you will end up with 4 splits for the cases that are `if`, `for`, `while`, statements with conditions and 2 paths for every other control-flow edge, like entering the first basic block from the Entry CFG node, or following a `break`, `continue` or `goto` (unconditional) CFG edge.
Documented this now in c2bd657065e0edcac742941c1a535444a90b1638, and added a unit test for the case when both callbacks are defined in 910abb384af9c8d51f32d2109a73ab900682957e.

> > What is the relationship of this new callback with the BranchCondition callback
> 
> +1, I am also interested to learn what is the main motivation behind the new callback. Do you have some example use cases in mind? I think users might get confused which one to use.

My motivation is to drive enabling or disabling an internal checker by entering or leaving a certain CFG block.
Concretely, if a short-circuiting operator leads to the guarded CFG block (aka. the true branch for the `&&` operator and the false branch for the `||` operator), enable a checker to detect certain constructs in a guarded block. And once we are leaving the guarded block disable the checker. I hope it clarifies the motivation and demonstrates the usefulness of such a callback.

I believe that there could be other similar use cases of this.

balazs-benics-sonarsource commented:
I know it didn't actually pass a week for a ping, but let me know if its on the horizon.

NagyDonat reviewed (commented):
I finished reviewing the non-test code and overall it LGTM, but I added two minor questions as inline comments. ~~I'll review the tests tomorrow.~~ Edit: I reviewed them today.

balazs-benics-sonarsource commented on clang/include/clang/StaticAnalyzer/Core/Checker.h:
```diff
@@ -548,6 +564,8 @@ class CheckerProgramPointTag : public SimpleProgramPointTag {
 template <typename CHECK1, typename... CHECKs>
 class Checker : public CHECK1, public CHECKs..., public CheckerBase {
 public:
+  using BlockEntrance = clang::BlockEntrance;
```
Yes, the class `check::BlockEntrance` would have priority within a checker, e.g. the DivByZero checker when spelling an unqualified name `BlockEntrance`. This means that the function parameter of that callback would refer to the wrong class: `checkBlockEntrance(const BlockEntrance &Entrance,...)`.
To circumvent this and provide the expected lookup rules, I need to force it to shadow this name with the `clang::BlockEntrance`, hence this using declaration here.
It's not pretty, but smart and has a low footprint. No users should know why it works.

NagyDonat replied:
> As I was thinking about this now, I think I'd prefer landing this first.
I'd need to backport this into clang-20, and that would make it (marginally) more complicated.

Sorry, I noticed this comment only when I already merged the checker family commit...

steakhal replied:
Dont worry.

balazs-benics-sonarsource committed:
NFC Move the check::BlockEntrance below check::Bind

NagyDonat commented on clang/unittests/StaticAnalyzer/BlockEntranceCallbackTest.cpp:
```diff
@@ -0,0 +1,357 @@
+    const auto *FD =
+        cast<FunctionDecl>(C.getLocationContext()->getStackFrame()->getDecl());
+
+    std::string Buffer =
+        (llvm::Twine("Within '") + FD->getIdentifier()->getName() +
+         "': branch condition '")
+            .str();
+    llvm::raw_string_ostream OS(Buffer);
+    Condition->printPretty(OS, /*Helper=*/nullptr,
+                           C.getASTContext().getPrintingPolicy());
+    OS << "'";
```
Bit sad to see this string juggling, but I understand that this is sometimes unavoidable...

No action needed.

NagyDonat commented on clang/unittests/StaticAnalyzer/BlockEntranceCallbackTest.cpp:
```diff
@@ -0,0 +1,357 @@
+                            AnalyzerOptions &AnOpts) {
+  AnOpts.CheckersAndPackages.emplace_back("test.BlockEntranceTester", true);
+  AnalysisConsumer.AddCheckerRegistrationFn([](CheckerRegistry &Registry) {
+    Registry.addChecker(&registerChecker<BlockEntranceCallbackTester>,
+                        &shouldAlwaysRegister, "test.BlockEntranceTester",
+                        "EmptyDescription", "EmptyDocsUri",
+                        /*IsHidden=*/false);
```
This will be yet another code fragment that will be prettified when I reintroduce `addMockChecker` in the spinoff NFC commit. There are already 10-15 tests that use `addChecker` calls with lots of dummy parameters like this, it will be nice to get rid of them :grin: 

No action expected.

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/CheckerDocumentation.cpp:
```diff
@@ -166,6 +179,23 @@ class CheckerDocumentation
+  ///
+  /// This callback can be used to obtain information about potential branching
+  /// points or any other constructs that involve traversing a CFG edge.
+  /// Note that when inlining a call, there is no CFG edge between the caller
+  /// and the callee. One will only see the edge between the entry block and
+  /// the body of the function once inlined.
```
Thanks for clarifying this! You're right that it's important to document this peculiarity of our graph (there is no BlockEdge from the caller to the callee + there is a BlockEdge from the imaginary entry block) and it would be very nice if you could squeeze something like this review comment into the source comment as well :smile:

NagyDonat reviewed (approved):
I also read the tests and the change LGTM if you move the explanations from the very helpful review comments https://github.com/llvm/llvm-project/pull/140924/files#r2107471659 and https://github.com/llvm/llvm-project/pull/140924/files#r2107491565 to source code comments.

In the tests I felt that it'd be a bit hard to decipher the meaning of the block identifiers like `B1` etc. -- but when I re-read the file I noticed that you included the very nice helper function `dumpCFGAndEgraph` (IIUC) for those who will need to debug broken cases in this test file :smile: Perhaps it would be even nicer if you included a commented out call to that function with "`// NOTE: Uncomment this if you want to decipher the meaning of 'B0', 'B1', ...`" to make its existence and role more obvious.

balazs-benics-sonarsource committed:
Merge branch 'main' into bb/upstream-check-block-entrance

balazs-benics-sonarsource committed:
Refine comments

balazs-benics-sonarsource committed:
Comment dumpCFGAndEgraph for debugging

balazs-benics-sonarsource committed:
Ease debugging

balazs-benics-sonarsource commented:
> In the tests I felt that it'd be a bit hard to decipher the meaning of the block identifiers like `B1` etc. -- but when I re-read the file I noticed that you included the very nice helper function `dumpCFGAndEgraph` (IIUC) for those who will need to debug broken cases in this test file 😄 Perhaps it would be even nicer if you included a commented out call to that function with "`// NOTE: Uncomment this if you want to decipher the meaning of 'B0', 'B1', ...`" to make its existence and role more obvious.

Changed the debugging interface a bit, and added comments explaining it in f339eb1dc8216da013d5a92f01332b43e0a75790

NagyDonat reviewed (commented):
@steakhal Thanks for the updates, I'm completely satisfied with them.

I don't see any connection between this commit and the buildbot failures :thinking: ... they are probably unrelated.

balazs-benics-sonarsource commented:
> @steakhal Thanks for the updates, I'm completely satisfied with them.
> 
> I don't see any connection between this commit and the buildbot failures 🤔 ... they are probably unrelated.

About 90% of the time they are unrelated. I don't usually put a confirmation to these messages.
